### PR TITLE
Add pre-commit support

### PR DESCRIPTION
**Pull request by abejgonzalez**
See https://github.com/firesim/firesim/pull/1329 for original PR.

This PR adds pre-commit support (https://pre-commit.com/) to run pre-commit hooks. Look at the `.pre-commit-config.yaml` file to see how pre-commit hooks are enabled. For a list of other supported plugins see https://pre-commit.com/hooks.html. To enable the hooks you have to run pre-commit install so I added that command to the `env.sh` file (this adds a small lag - ~O(s) - to the 1st time the `env.sh` is invoked but afterwards is a noop) (see `build-setup.sh`). Additionally, to run the hooks on all files in the repo I ran `pre-commit run --all-files`. Normally only the files that are committed are checked with the pre-commit hooks.

**Related PRs / Issues**:

**Type of change**:
- [ ] Bug fix
- [x] New feature
- [ ] Other enhancement

**Impact**:
- [ ] RTL change
- [ ] Software change (RISC-V software)
- [ ] Build system change
- [x] Other

**Contributor Checklist**:
- [x] Did you set `main` as the base branch?
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you state the type-of-change/impact?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you mark the PR with a `changelog:` label?
- [x] (If applicable) Did you update the conda `.conda-lock.yml` file if you updated the conda requirements file?
- [x] (If applicable) Did you add documentation for the feature?
- [x] (If applicable) Did you add a test demonstrating the PR?

- [x] (If applicable) Did you mark the PR as `Please Backport`?

### CODE REVIEW

**Comment by jerryz123** [comment]
Neat. What happens when you break one of the rules? Does it fix your commit for you?

**Comment by tymcauley** [comment]
It looks like it edited some `.svg` files. Is that bad? Seems like it shouldn't search for whitespace issues in a binary file.

**Comment by jerryz123** [comment]
SVGs are XML files, interestingly.

**Comment by tymcauley** [comment]
> SVGs are XML files, interestingly.

🤯

**Comment by abejgonzalez** [comment]
If there is a breakage it can either warn you and exit or edit the files itself.

**Comment by abejgonzalez** [comment]
I also just enabled all the pre-commit hooks since I think there isn't much movement in the repo right now. So now could be a good time to just force formatting on everything.

**Comment by abejgonzalez** [comment]
For reference on the Clang formatting: https://zed0.co.uk/clang-format-configurator/

**Comment by abejgonzalez** [comment]
Note that the documentation error is not due to this PR (the BOOM documentation / website is down for some reason)

File: vlsi/example-vlsi [hunk 1]
```diff
@@ -10,53 +10,59 @@ from hammer_vlsi import CLIDriver, HammerToolHookAction
...
+
 class ExampleDriver(CLIDriver):
     def get_extra_par_hooks(self) -> List[HammerToolHookAction]:
         extra_hooks = [
             # Default set of steps can be found in the CAD tool plugin's __init__.py
-
             # make_pre_insertion_hook will execute the custom hook before the specified step
             # SYNTAX: make_pre_insertion_hook("EXISTING_STEP", INSERTED_HOOK)
             # hammer_vlsi.HammerTool.make_pre_insertion_hook("route_design", example_add_fillers),
-
```
Comment (jerryz123) [inline vlsi/example-vlsi:47]: I don't think these changes are an improvement.

File: vlsi/view_gds.py [hunk 2]
```diff
@@ -12,70 +12,70 @@
...
-            #(20, 0),    #M2
-            (20, 251),  #M2 lbl
-            #(25, 0),    #V2
-            #(30, 0),    #M3
-            (30, 251),  #M3 lbl
-            #(35, 0),    #V3
-            #(40, 0),    #M4
-            (40, 251),  #M4 lbl
-            (45, 0),    #V4
-            (50, 0),    #M5
```
Comment (jerryz123) [inline vlsi/view_gds.py:54]: Another example, this destroyed the formatting

File: fpga/src/main/resources/vcu118/sdboot/include/devices/gpio.h [hunk 3]
```diff
@@ -3,22 +3,22 @@
...
-#define GPIO_FALL_IE    (0x20)
-#define GPIO_FALL_IP    (0x24)
-#define GPIO_HIGH_IE    (0x28)
-#define GPIO_HIGH_IP    (0x2C)
-#define GPIO_LOW_IE     (0x30)
-#define GPIO_LOW_IP     (0x34)
-#define GPIO_IOF_EN     (0x38)
-#define GPIO_IOF_SEL    (0x3C)
-#define GPIO_OUTPUT_XOR    (0x40)
+#define GPIO_PULLUP_EN (0x10)
```
Comment (jerryz123) [inline fpga/src/main/resources/vcu118/sdboot/include/devices/gpio.h:10]: Not an improvement

**Review by jerryz123 — CHANGES_REQUESTED** [review]
I don't like forcing formatting on source code like this. No-trailing-whitespace is fine, but restructuring code and comments like this is too heavy-handed.

**Comment by abejgonzalez** [comment]
I personally think having a linter/code-formatter is better than not having one at all, but we can push that to a later discussion. For now this PR will just focus on whitespace management + checking files.

**Comment by abejgonzalez** [comment]
@jerryz123 This is now ready for rereview

**Review by jerryz123 — APPROVED** [review]
Thanks for fixing all the trailing-whitespace/newline issues. The git hooks thing seems useful.

**Pull request merged**